AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -2049,6 +2052,67 @@ - (BOOL)supportsRangeManagedInterfaceState
   return ASHierarchyStateIncludesRangeManaged(_hierarchyState);
 }
 
+- (id<ASTraitCollection>)traitCollection
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (_traitCollection == nil) {
+    _traitCollection = [[ASDisplayTraitCollection alloc] initWithTraits:_internalTraits];
+  }
+  return _traitCollection;
+}
+
+- (void)setTraitCollection:(id<ASTraitCollection>)traitCollection
```

Comment:
Is setting a trait collection something that we should do? `UITraitEnvironment` exposes `traitCollection` as readonly. 

Instead I wonder if the default behavior should be to determine the traits based upon the current state of the node. Can we try to iterate up the responder chain calling `traitCollection`? If a subclass overrides `traitCollection` to always be a particular set of traits, that would stop the responder iteration and propagate the custom trait collection back to this node. If we reach the top of the responder chain, could we make an `ASTraitCollection` based on that node's `UITraitCollection` (not sure how this will work if the top responder is not view-backed)? If not then we'd need all the nasty logic to determine size classes.